File: javascript/day3.js
var { inputDay3 } = require('./input');
var assert = require('assert');

function isTree(area, coordinate){
    let [col, row] = coordinate;
    let position = area[row][col];

    return position == "#";
}

function shouldExpandMap(area, coordinate, slope=[1,1]){
    let currentIndex = coordinate[0];
    let maxIndex     = area[0].length - 1;
    let rightSteps   = slope[0];

    return ((maxIndex - currentIndex) < rightSteps);
}

function expandMap(currentArea, addingArea){
    assert(
        currentArea.length == addingArea.length,
        "Adding area doesn't have the same shape as the current area"
        );

    // Note: map function works very differently in javascript compared to python,
    // so you can't use map function as following
    // let newArea = currentArea.map(function (a, b){return a + b}, addingArea);
    // The working way is like this:
    let newArea = currentArea.map(function(a, i){return a + addingArea[i]})
    return newArea
}

function makeAMove(position, slope=[1,1]){
    let [x, y] = position;
    return [x+slope[0], y+slope[1]]
}

// PART 1
function countTreesOnPath(area, slope=[3,1]){
    let movingPosition = [0,0];
    let fullMap = area;
    
    if (isTree(fullMap, movingPosition)){
        var countTrees = 1;
    } else {
        var countTrees = 0;
    }

    while (movingPosition[1] < (fullMap.length - slope[1])){
        if (shouldExpandMap(fullMap, movingPosition, slope)){
            fullMap = expandMap(fullMap, area);
        }

        movingPosition = makeAMove(movingPosition, slope);

        if (isTree(fullMap, movingPosition)){
            countTrees += 1;
        }
    }

    return countTrees
}

console.log(countTreesOnPath(inputDay3, [3,1]));

// PART 2
let slopes = [[1,1],[3,1],[5,1],[7,1],[1,2]];
let treeOnEachSlope = slopes.map(function(a){return countTreesOnPath(inputDay3, a)});

console.log(treeOnEachSlope.reduce(function(a,b){return a*b}));

File: python/day3.py
from functools import reduce

with open("/home/patrick/Documents/projects/1problem-3languages/day3_input.txt", "r") as f:
    file_content = f.readlines()

my_map = list(map(lambda x: x.replace("\n", ""), file_content))

def is_a_tree(area, coordinate):
    col = coordinate[0]
    row = coordinate[1]
    position = area[row][col]

    return "#" == position

def make_a_move(coordinate=[0,0], slope=[3,1]):
    x = coordinate[0]
    y = coordinate[1]
    step_right = slope[0]
    step_down = slope[1]

    return [x+step_right, y+step_down]

def expand_map(current_area, adding_area):
    assert len(current_area) == len(adding_area), \
           "Adding area doesn't have the same length as of current area"

    new_area = list(map(lambda a, b: a + b, current_area, adding_area))
    return new_area

def should_expand_map(current_area, coordinate):
    current_index = coordinate[0]
    max_index     = len(current_area[0]) - 1
    right_steps   = coordinate[0]

    return ((max_index - current_index) < right_steps)

# PART 1
def count_trees_on_path(area, slope=[3,1]):
    position     = [0,0]
    count_trees  = 0
    current_area = area

    while position[1] < (len(area) - slope[1]):
        if is_a_tree(current_area, position):
            count_trees += 1

        if should_expand_map(current_area, position):
            current_area = expand_map(current_area, area)

        position = make_a_move(position, slope)

    return count_trees

print(count_trees_on_path(my_map))

# PART 2
slopes = [[1,1], [3,1], [5,1], [7,1], [1,2]]
tree_counts = list(map(lambda x: count_trees_on_path(my_map, x), slopes))

print(reduce(lambda x, y: x*y, tree_counts))